apple/swift-docc
# Compute beta platform information for non-symbol documentation

Bug/issue #, if applicable: rdar://129355087

## Summary

When configuring availability information for an article using the `@Available` directive, beta tags are never shown/computed even when the platforms are configured as beta:

Markdown:
```
# Sample Code Article

@Metadata {
    @Available(iOS, introduced: "18.0")
    @Available(Xcode, introduced: "16.0")
    @Available(macOS, introduced: "15.0")
    @Available(visionOS, introduced: "2.0")
    @SupportedLanguage(swift)
}
```

DocC invocation:
```
docc preview SampleCode.docc \
--platform name=macOS,version=15.0.0,beta=true \
--platform name=iOS,version=18.0.0,beta=true \
--platform name=visionOS,version=2.0.0,beta=true \
--platform name=Xcode,version=16.0.0,beta=true
```

Preview:
<img width="384" alt="Screenshot 2024-06-19 at 3 35 04 PM" src="https://github.com/apple/swift-docc/assets/15234535/da66bb94-0621-4511-83fb-2209aba71577">

I would expect the platforms to be marked with a beta badge, similarly to how that happens for symbols:
<img width="395" alt="Screenshot 2024-06-04 at 16 02 34" src="https://github.com/apple/swift-docc/assets/15234535/051c2cab-66db-42a8-a28f-07656bb781ab">

### Fix

This bug was being caused by the [`AvailabilityRenderItem.isBeta`](https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Model/Rendering/Symbol/AvailabilityRenderMetadataItem.swift#L87-L88) property (which is used to render the beta badges) never being set for the case when availability information comes from the `@Available` directive.

It was being set for the symbol graph case correctly:
https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Model/Rendering/Symbol/AvailabilityRenderMetadataItem.swift#L125-L141

But not for the metadata case:
https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Model/Rendering/Symbol/AvailabilityRenderMetadataItem.swift#L143-L150

I've made it so that it's always set, which required a minor rework of the `@Available` directive.

### Changes

- The `@Available` directive now parses the [`introduced`](https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Semantics/Metadata/Availability.swift#L93-L95) field as a [`VersionTriplet`](https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Infrastructure/External%20Data/ExternalMetadata.swift#L14), instead of as a `String`. This is useful for:
         - Validating the user input is in a valid format
         - To be able to compare the availability from the `@Available`  directive with the platform metadata from DocC's context, which is then used to determine whether a platform has been configured as being in beta.
- [`.stringRepresentation(precisionUpToNonsignificant:)`](https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Model/Rendering/Symbol/AvailabilityRenderMetadataItem.swift#L28-L46) has been moved to [`VersionTriplet`](https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Infrastructure/External%20Data/ExternalMetadata.swift#L14). This is a non-source-breaking change as this was internal API.
- Sets [`AvailabilityRenderItem.isBeta`](https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Model/Rendering/Symbol/AvailabilityRenderMetadataItem.swift#L87-L88) when [`AvailabilityRenderItem.init?(_ :current:)`](https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Model/Rendering/Symbol/AvailabilityRenderMetadataItem.swift#L143) is called.
- Makes [`AvailabilityRenderItem.isBeta`](https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Model/Rendering/Symbol/AvailabilityRenderMetadataItem.swift#L87-L88), [`AvailabilityRenderItem.introduced`](https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Model/Rendering/Symbol/AvailabilityRenderMetadataItem.swift#L63-L64) and [`AvailabilityRenderItem.name`](https://github.com/apple/swift-docc/blob/f53b6e03cfc99d4e684d54351a8cec7fe36caabc/Sources/SwiftDocC/Model/Rendering/Symbol/AvailabilityRenderMetadataItem.swift#L60-L61)  into `let` properties as they are integral to availability rendering. This is to avoid any future bugs, as the code will now no longer compile unless this property is set in an initializer.
- Adds diagnostic explanation for non-convertible directive arguments, to be emitted as part of the diagnostic explanation whenever the author provides an argument which is not convertible to the expected type (https://github.com/apple/swift-docc/pull/959/commits/198c827829bdcafcf9b2b275dfc3a9545c819216). 

## Dependencies

N/A

## Testing

Steps:
1. Download the example documentation bundle: 
[SampleCode.docc.zip](https://github.com/user-attachments/files/15902327/SampleCode.docc.zip)
2. Preview the documentation:
```
docc preview SampleCode.docc \
--platform name=macOS,version=15.0.0,beta=true \
--platform name=iOS,version=18.0.0,beta=true \
--platform name=visionOS,version=2.0.0,beta=true \
--platform name=Xcode,version=16.0.0,beta=true
```
3. Verify that all platforms as showing as being in beta, as well as the article itself.

More generally, this can be verified by adding availability information to any documentation markdown file (excepting a symbol extension file):
```
@Metadata {
    @Available(iOS, introduced: "2.0")
}
```
And defining that platform version as beta using docc's `--platform` flag.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.


### Verification

**Unit testing**
Modified existing tests for `VersionTriplet` and `SymbolGraph.SemanticVersion`.
Added unit tests to `MetadataAvailabilityTests` and `PlatformAvailabilityTests`.

**Site rendering**
Before                                  |  After
:-------------------------:|:-------------------------:
<img width="1172" alt="Screenshot 2024-06-19 at 3 29 51 PM" src="https://github.com/apple/swift-docc/assets/15234535/cf8a660a-d936-4765-9eab-44b512593685"> | <img width="1175" alt="Screenshot 2024-06-19 at 3 28 05 PM" src="https://github.com/apple/swift-docc/assets/15234535/3209b4e9-c98d-4b95-a681-a692ca9009f4">

**Documentation**

Updated documentation for `@Available` directive.

**Error cases**

When an invalid semantic version number is passed to the `@Available` directed, the following warning diagnostic is emitted:
```
warning: Cannot convert 'hello' to type 'VersionTriplet'
Available expects an argument for the 'introduced' parameter that's convertible to a semantic version number ('[0-9]+(.[0-9]+)?(.[0-9]+)?')
```
